### ISSUE TITLE
Initialisation of overridden or excluded defs and vaes

### ISSUE THREAD

**Issue by kjx** [gracelang/language#54]
What happens when 

   Class Top {
       def X = 3
    }

    Class Bot {
        Inherit Top
           Excludes X 
        Method X { }
    }

- where does the init code go?
- what if there's no overriding method declaration?

**Comment by KimBruce** [comment]
In languages with facilities like this, “excludes” usually means hiding.  For example, if Top has a method m that uses x, then x must stay around to be used by the method (of course this is even more important if x is a variable rather than a def).

In my opinion, an excludes statement really means the item becomes “private”. I.e., it’s around is subclasses, but not visible there. I’d like alias to be the same (though at the place defined, rather than by a subclass excluding it).

Kim

**Comment by kjx** [comment]
> In languages with facilities like this, “excludes” usually means hiding

right - and perhaps that's what we do here

> For example, if Top has a method m that uses x, then x must stay around to be used by the method (of course this is even more important if x is a variable rather than a def).

So my Pomona draft said:

> for every declaration that IS excluded, add an abstract declaration to the bag
>          (multiple abstract declarations of the same name are coalesced) 

(perhaps _excluded_ would be a better name than _abstract_, but you get the idea)

Andrew's semantics at #38 just say the name is removed from the bag. 

I'm pretty sure we said weren't doing Eiffel _deep renaming_: there's only one binding for _x_ in the object, and every self request will resolve to that _x_. This is why we decided not to have private. 

> I’d like alias to be the same

If alias is the same, then once again we have private, we have more complicated request resolution, and, it's more difficult for a class to hand down access to an overridden definition to a subclass.

**Comment by apblack** [comment]
I think that under the current inheritance semantics, we have to execute the initialization code of `top` when it is inherited.  Here the code is just `3`, but it might be

```grace
class top {
   def x = valueOf { print "initializing x"; 3 }
}
```

So even though `x` is excluded from `bot`, we still have to execute the initialization.

At the risk of sounding like a broken record, I don't think that the current initialization semantics makes sense.  Nevertheless, until we change it, the meaning of this example seems clear.

> where does the init code go?

I don't think that I understand this question. The code _goes_ into memory somewhere.  The initialization code of `top` is _executed_ as part of the effect of initializing `bot`.

> what if there's no overriding method declaration (for `x`, presumably?)

Then there is no `x` method in the object.   jkx, I think that you want the compiler to cons-up a new method body equivalent to

    method x { 
        TraitCompositionException.raise "method x was excluded but is nevertheless requested"
    }

Is that right?   I'm opposed to that, as your overworked implementor.  If the method is excluded, then the user will get a `NoSuchMethod` exception instead, which is only marginally less helpful.

**Comment by apblack** [comment]
KimBruce, I have some sympathy for the point of view that `x` ought to be visible to the methods of the trait that declared it, but not elsewhere.  But allowing this would complicate our object model enormously: we would need to introduce lexically bound methods (aka procedures) and a notation for calling them, neither of which do we have.  Moreover, what was originally written as a method request would become lexically-bound after the fact.  And what if `top` were not a top-level trait, but was defined as a composition of a number of smaller traits?  The components might use `x`.  Then`x` is "hidden".  How do those other trait methods access `x`?

I think that trying to some up with something coherent here will lead us to C++.  **Let's not go there.**
An excluded method is _not there_, and requesting it will give rise to a `NoSuchMethod` error, unless it is later provided.

**Comment by kjx** [comment]
> An excluded method is not there, and requesting it will give rise to a NoSuchMethod error,

I can live with that, or with them converting to abstract/required methods.  Kim, comments?

if/when we agree, we should remember NOT to close this, but to move it from now to resolved to be specd.

**Comment by KimBruce** [comment]
I think it is the wrong semantics as it breaks code that the implementor may not even be aware of (that is being inherited from the superclass), requiring a deeper understanding of the superclass than is normally warranted, though the issues are at least related to that of overriding a method when you don’t know how it was used in the superclass (though you can at least try to obey invariants).

Nevertheless, it’s an exotic enough feature that I’ll likely never use or teach (and I wish we actually didn’t have it in the language!!).  So, do as you like.

Kim

**Comment by apblack** [comment]
It seems to me that we have **agreed** that an excluded method is not there, and requesting it will give rise to a `NoSuchMethod error`.  

As Kim says, this can be dangerous!  If you exclude a method, you has better know what you are doing!  For example, it may be an interface method that isn't used externally and you don't want clients to have access to.  I'm happy to have cautionary language in the spec, although, long-term, we are going to need a "users guide" as well as a spec, and that is probably the right place for such a discussion.

**Comment by kjx** [comment]
fine with me - moving to be spec'd.

and yes: a spec is a spec, a user guide is a user guide.
there are already some discussions in the spec in places we could move out.

**Comment by kjx** [comment]
drafted - https://github.com/gracelang/language/blob/portland-james/spec.md#combination-and-initialisation

**Comment by KimBruce** [comment]
I don't see where the spec actually says whether the "excluded" features are still in the object (and accessible by code from the super class.  E.g., is the following legal (I think it should be).

    class c {
         var x:= = 3
         method m {x+2}
    }

    class d {
        inherits c
            excludes x
        ...
     }
     var x := 47
     d.m

In my mind the call of d.m should be legal, even though x is not accessible in any of the code in class d, and the accessed value of x should be the one specified in c, not the global x.

**Comment by apblack** [comment]
James and I discussed essentially the same example on Saturday.  We agreed that the `x` in the `x+2` in the c's method m is a self request on the `var x` defined in `c`.  It's meaning does not change just because c is inherited into a context, such as `d`,  where there is no `self.x`.

So, as written, `d.m` gives ride to a `NoSuchMethod` error.

But I can also use `d` as follows:

```grace
class e {
    inherits d
    def x = -2
}
```

and now requests of `e.m` will request `d.x`, and `m` will answer 0.

**Comment by kjx** [comment]
> I don't see where the spec actually says whether the "excluded" features are still in the object 

It says excluded attributes are excluded. We don't even give a special "you tried to call an excluded feature" error - they are just gone. 

> (and accessible by code from the super class. E.g., is the following legal (I think it should be).
> class c {
>      var x:= = 3
>      method m {x+2}
> }
> 
> class d {
>     inherits c
>         excludes x
>     ...
>  }
>  var x := 47
>  d.m
I believe this is "legal" As in not being a trait conflict. It will break at runtime. Presumably it could also be caught by a type checker although we don't know yet. 
> In my mind the call of d.m should be legal, 
The call is legal. 
Then the method should  crash with a request not understood error. 

> even though x is not accessible in any of the code in class d, 
Right its not. 
> and the accessed value of x should be the one specified in c, not the global x.
Yes it should but d's "view of c" doesn't have a x.   The spec should now say x in c resolves as a self call. 
That self call will fail as I've said above. 

Ever since Pomona we agreed we were doing **shallow** aliasing in a single namespace.  We would alias  declarations but never look inside method bodies. Each object would have only one namespace and all self-requests are resolved In that namespace. 

There are small changes that would help a bit: if exclude turns a method abstract then it's clearer what will happen I think. 

But if you're really unhappy about this  I'm not sure what to suggest apart from pushing all of traits out to the next version of the spec. 

J

**Comment by KimBruce** [comment]
Let’s talk about this tomorrow.  I think we’ve seen lots of examples that show that there are lots of problems with this interpretation of exclude.  I don’t think we need to be as drastic as to eliminate traits from the spec.  However, it might be easier to just put of “excludes” if we can’t come to an agreement.

Kim

**Comment by kjx** [comment]
> However, it might be easier to just put of “excludes” if we can’t come to an agreement.

yep. and in the spirit of simplification I'm inclined towards that too. 
If local abstract declarations override inherited methods  (see #44)
then it seems that

```
object { 
   use someTrait
      exclude foo(_) 
}
```
can be replaced with 
```
trait someTraitWithoutFoo {  
  use someTrait
  method foo is abstract { }
}

object {
  use someTraitWithoutFoo {
}
```

If Grace traits are primarily for combining dialects and organising libraries we write (planned reuse) then I think we could live with this: we don't lose any power, just a bit of concision.  Alias cannot be removed in the same way, precisely because of the "one namespace per object" rule -- and they also support another way of resolving the trait conflict: alias and then redefine the method you want,   rather than excluding the methods you don't.  

```
object {
      use SomeTraitWithFoo //we don't want that foo
      use SomeTraitAlsoWithFoo// we want this one
        alias foo(_) = fooWeWant(_)
     method foo(x) = { fooWeWant(_) } 
 }
```

**Comment by kjx** [comment]
agreed - exclude effectively makes things `required`

**Comment by apblack** [comment]
The resolution of this (on 8th March 2016) was that excluding a method should have thesame effect as overriding the existing method with a *required* method.
